src/data/articles/scala-3-the-evolution-of-a-macro.mdx

```diff
@@ -0,0 +1,1829 @@
+---
+author: daniel-beskin
+category: guide
+difficulty: advanced
+excerpt: Discover how a macro can evolve to reduce boilerplate in your code
+publishedDate: 2025-04-10
+repositoryUrl: https://github.com/ncreep/scala3-macro-tracing-blog
+tags: [fp, scala, scala-3, macros]
+title: "Scala 3: The Evolution of a Macro"
+---
+## Introduction
+
+In this post we are going to explore how Scala 3's macros can be used to simplify the use of a semi-realistic and boilerplaty tracing library.
+We'll see how we can reduce boilerplate using different levels of macro implementations, starting from simple but minimally helpful, and evolving our macros into something more complex but almost magical.
+
+Each step in this evolution will use various macro capabilities and we'll see how to leverage them to reduce boilerplate in our code.
+
+## Disclaimer
+
+This post is long, going through a lot of code. Please take your time reading it, and let each part sink in before moving on.
+
+Throughout the code examples basic understanding of Scala 3's macro APIs is assumed. Things like quotes/splices, `Expr`, `Tree`, `Symbol`, etc., should be familiar to the reader. You can think of this blog as a supplement to Rock the JVM's [metaprogramming course](/courses/scala-macros-and-metaprogramming). For an in-depth introduction to macros you can read [this](scala-3-macros-comprehensive-guide) article.
+
+## Setup
+
+All the code from the post is available in <a href={frontmatter.repositoryUrl}>this</a> repo. The project is set up using sbt. If you want to follow along in a REPL, please use the project's sbt console:
```

Comment:
use markdown links

Reply:
You mean drop the `repositoryUrl` from the front matter and link directly? Or is there some way to reference variables from the front matter directly from plain markdown syntax (I couldn't get it to work)?

Reply:
I saw you used Markdown links for non-variables.
Not a big deal.